src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,220 @@
+    return;
+}
+
+static void tcp_many(Error error,
+                     Var<Entry> entry,
+                     std::map<std::string, std::vector<std::string>> fb_service_ips,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Var<Entry>> cb
+                    ) {
+    logger->info("starting tcp_many");
+    if (error) {
+        cb(entry);
+        return;
+    }
+
+    // if we find any blocked TCP later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int ips_count = 0;
+    for (auto const& service_and_ips : fb_service_ips) {
+        // skipping stun for now
+        if (service_and_ips.first == "stun") {
+            continue;
+        }
+        for (auto const& ip : service_and_ips.second) {
+            ips_count += 1;
+        }
+    }
+    Var<int> ips_tested(new int(0));
+    if (ips_count == *ips_tested) {
+        cb(entry);
+        return;
+    }
+
+    auto tcp_cb = [=](std::string service, std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            if (!!err) {
+                logger->info("tcp failure to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = true;
+                (*entry)["facebook_dns_blocking"] = true;
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = false;
+            }
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
+                if (close_txp == true) {
+                    txp->close([=] {
+                        cb(entry);
+                        return;
+                    });
+                } else {
+                    cb(entry);
+                    return;
+                }
+            } else {
+                if (close_txp == true) {
+                    // XXX optimistic closure
+                    txp->close([=] {});
+                }
+            }
+        };
+    };
+
+    for (auto const& service_and_ips : fb_service_ips) {
+        std::string service = service_and_ips.first;
+        // skipping stun for now
+        if (service == "stun") {
+            continue;
+        }
+        for (auto const& ip : service_and_ips.second) {
+            Settings tcp_options;
+            tcp_options["host"] = ip;
+            int port = 443; //XXX hardcoded
+            tcp_options["port"] = port;
+            tcp_options["net/timeout"] = 10.0; //XXX hardcoded
+            templates::tcp_connect(tcp_options, tcp_cb(service, ip, port),
+                                   reactor, logger);
+        }
+    }
+    return;
+}
+
+void facebook_messenger(std::string input, Settings options,
```

Comment:
This `options` variable is the one that you should pass around and fetch options from.

Reply:
done